### PR TITLE
Final LibGit2 credential callback changes

### LINKED ISSUE #18403: Fix SSH tests so they pass on buildbots, and re-enable them
ref #17651 and #18397 - need to figure out what was wrong here and fix it so we can turn these tests back on without blocking nightlies for a month again.
#18306 is closely related

### PR DESCRIPTION
**Pull request by omus**
Should be the last round of bugfixes and minor features to the LibGit2 credential callback system before I move onto integrating git credential helpers. There are quite a few small changes in this PR including:

* `CredentialPayload`'s credential field is now used a temporary storage. We can safely do this without accidentally modifying cached credentials because of #23711
* Load credential information from ENV only once. Allows future iterations to use user input as defaults
* Created new `isfilled` function which verifies a credential is ready to be used in authentication
* Introduced `modified` check which allows us to abort when all authentication methods have failed
* Re-prompting for an SSH key will always ask for the private key to allow the user to change it
* Introduced a way to prompt for the public key when set to an invalid file
* All previously broken/commented out callback tests are now enabled
* Limit number of user prompts to 3
* Support ~ in SSH prompts for private key and public key

Additionally, at this point I think the `credential_loop` tests cover our testing of the SSH code paths. The ancient "SSH" testset has been removed which fixes #18403, #18306.

Part of #20725.

### CODE REVIEW

**Comment by omus** [comment]
Interesting, this code is no longer broken on Windows:
```julia
if Sys.iswindows() && LibGit2.version() >= v"0.26.0"
    # see #22681 and https://github.com/libgit2/libgit2/pull/4055
    @test_broken ex.code == LibGit2.Error.EAUTH
    @test ex.code == LibGit2.Error.ERROR
else
    @test ex.code == LibGit2.Error.EAUTH
end
```
I believe this may have been caused by the changes in ["Set error message when auth fails"](https://github.com/JuliaLang/julia/pull/23770/commits/f5065b36cbbfe01e8cb5b287fe709d53fe9882d1) which now always sets the error message. It's possible that the LibGit2 v0.26.0 update would modify the error code if no new error message was set (working theory).

File: base/libgit2/callbacks.jl [hunk 1]
```diff
@@ -42,20 +42,38 @@ end
...
+function exhausted_abort()
+    ccall((:giterr_set_str, :libgit2), Void,
+          (Cint, Cstring), Cint(Error.Callback),
+          "All authentication methods have failed.")
+    return Cint(Error.EAUTH)
+end
+
 function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload, username_ptr)
     creds = Base.get(p.credential)::SSHCredentials
+    modified = false
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: Some of these `modified` changes seem like they could simplified (e.g., You could probably just have `modified = p.first_pass && isfilled(creds)` at the beginning).

File: base/libgit2/callbacks.jl [hunk 2]
```diff
@@ -150,40 +183,46 @@ end
...
-                isnull(response) && return user_abort()
-                userpass = unsafe_get(response)
-                isempty(userpass) && return user_abort()  # Ambiguous if EOF or newline
-            end
+    if p.remaining_prompts > 0 && (!modified || !isfilled(creds))
+        prompt_url = git_url(scheme=p.scheme, host=p.host)
+        if Sys.iswindows()
+            response = Base.winprompt(
+                "Please enter your credentials for '$prompt_url'", "Credentials required",
+                isempty(creds.user) ? p.username : creds.user; prompt_username=true)
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: I'd be inclined to move the ternary out of the if-block and put the remaining code into an `user_auth_prompt` function.

File: base/libgit2/callbacks.jl [hunk 3]
```diff
@@ -65,82 +83,97 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
...
             isnull(response) && return user_abort()
-            publickey = unsafe_get(response)
+            creds.pubkey = expanduser(unsafe_get(response))
         end
 
-        if isempty(passphrase) && is_passphrase_required(privatekey)
+        # Ask for a passphrase when the private key exists and requires a passphrase
+        if isempty(creds.pass) && is_passphrase_required(creds.prvkey)
             if Sys.iswindows()
                 response = Base.winprompt(
```
Comment (rofinn) [inline base/libgit2/callbacks.jl]: Move prompt logic into `ssh_auth_prompt`?

**Review by rofinn — APPROVED** [review]
I don't see any obvious logical issues with the code, just some style and organizational suggestions.

**Comment by omus** [comment]
Ok, I worked my way through the AppVeyor issue. As it turns out the commit ["Error when credentials are not modified"](https://github.com/JuliaLang/julia/pull/23770/commits/05fb5c3940685c9a303b4e3debd74c847554cf19) corrects the broken test. In this commit we throw a EAUTH error from the callback when the credentials are not filled. I'll update the libgit2-online tests to reflect this but I also think it could be reasonable to remove the "with empty credentials" test.

File: base/libgit2/callbacks.jl (hunk 1 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: It may be better if I drop `modified` and just do a comparison between the initial credentials and the current credentials. I'll do some checks.

Comment (omus) [reply base/libgit2/callbacks.jl]: My idea of doing the comparison won't work. If a user were to enter the same invalid credentials twice in a row they would get an "All authentication methods have failed" error. The name `modified` isn't quite the right term as this just verifying that a new authentication method has been attempted.

File: base/libgit2/callbacks.jl [hunk 4]
```diff
@@ -65,82 +83,97 @@ function authenticate_ssh(libgit2credptr::Ptr{Ptr{Void}}, p::CredentialPayload,
...
-        prompt_url = git_url(scheme=p.scheme, host=p.host, username=username)
+        prompt_url = git_url(scheme=p.scheme, host=p.host, username=creds.user)
 
         # For SSH we need a private key location
-        if !isfile(privatekey)
+        last_private_key = creds.prvkey
+        if !isfile(creds.prvkey) || !modified
             response = Base.prompt("Private key location for '$prompt_url'",
-                default=privatekey)
+                default=creds.prvkey)
```
[the remark this answers is not in the document]
Comment (omus) [reply base/libgit2/callbacks.jl]: I've addressed this by renaming `prompt_url` to just `url` which means I can fit this keyword on a single line.

File: base/libgit2/callbacks.jl (hunk 1 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: I reworked this a little bit. I've ended up changing `modified` to `revised`. I think this term is better suited for the current behaviour. As for `revised = p.first_pass && isfilled(creds)` I've decided to keep the `if` block as it matches the way `revised` is used in the reset of the code.

File: base/libgit2/callbacks.jl (hunk 3 above)
Comment (omus) [reply base/libgit2/callbacks.jl]: Separating these into new functions hasn't turned out well. I've decided to stick with the status quo.

**Comment by omus** [comment]
During the investigation on changing how `modified` worked I discovered a corner case where if prompting was disabled, ssh agent had failed, and no ENV variables were set we would fail to try the default `~/.ssh/id_rsa` key. I've added a test to ensure that the default SSH key will be used without prompting.

**Comment by omus** [comment]
The PR should be ready to go now. I think this should be merged rather than squashed to keep the history understandable. I've run the LibGit2 tests locally on each commit.

**Comment by omus** [comment]
Found a minor issue with the LibGit2 library I'll patch yet.

**Comment by omus** [comment]
After digging into the LibGit2 issue the problem turned out to bug where the credential callback loop is aborted if the SSH passphrase is incorrect. Addressing this now makes `Pkg.update` much better as now you can accidentally type your SSH passphrase wrong and get re-prompted right away.

Additionally, after doing more experimenting with `Pkg.update` there was an issue where cached credentials were being ignored when the SSH agent was enabled. Now that these are all addressed `Pkg.update` authentication finally works the way I always wanted it to.
